src/collection/iteration.js

```diff
@@ -122,29 +122,34 @@ let elesfn = ({
 
 elesfn.each = elesfn.forEach;
 
-// implement an iterator only for those instances which support Symbol.iterator
-if (typeof Symbol != "undefined" && typeof Symbol.iterator != "undefined") { // eslint-disable-line no-undef
-  elesfn[Symbol.iterator] = function() { // eslint-disable-line no-undef
-    let entry = { value: undefined, done: false };
-    let i = 0;
-    let length = this.length;
-
-    return {
-      next: () => {
-        if ( i < length ) {
-          entry.value = this[i++];
-        } else {
-          entry.value = undefined;
-          entry.done = true;
+const defineSymbolIterator = () => {
+  const typeofUndef = typeof undefined;
+  const isIteratorSupported = typeof Symbol != typeofUndef && typeof Symbol.iterator != typeofUndef; // eslint-disable-line no-undef
+
+  if (isIteratorSupported) {
+    elesfn[Symbol.iterator] = function() { // eslint-disable-line no-undef
+      let entry = { value: undefined, done: false };
+      let i = 0;
+      let length = this.length;
+
+      return {
+        next: () => {
+          if ( i < length ) {
+            entry.value = this[i++];
+          } else {
+            entry.value = undefined;
+            entry.done = true;
+          }
+
+          return entry;
+        },
+        [Symbol.iterator]: function() { // eslint-disable-line no-undef
+          return this;
         }
```

Comment:
I can't think of a case where consumer code would have direct access to the object defined on line 135, so I'm not sure about the purpose of defining `[Symbol.iterator]` on that iterator object.  It probably does no harm, but I'm curious about the motivation for including it.

Reply:
I may be misunderstanding your question, but the consumer may provide access to the object with something like `let iter = eles[Symbol.iterator]()`. So then the purpose of defining a `[Symbol.iterator]` property on this object is solely to permit the consumer to be able to treat `iter` as an iterable. A potential (albeit somewhat contrived) use case might be that you want to partially iterate over a collection but still be able to resume iteration from where it left off in a separate loop. An example of this behaviour can be found in the updated test file.

As far as motivation goes, that simply came about after reading the advisory on the [iteration protocols MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) (right below the section "the iterator protocol" in the highlighted yellow rectangle), also mentioned [here](https://exploringjs.com/es6/ch_iteration.html#sec_implementing-iterables) (under section 21.5.1). It appears that built-in es6 iterators also follow this pattern. All that being said, even in my limited experience "because I read about it somewhere" intuitively seems like a bad justification for implementing something. So I'd be entirely comfortable with it if you took it out.

Reply:
OK, let's leave it in as long as it's undocumented.  That way, it's easy to publicly release that functionality in future with a documentation update.  Since it's undocumented, we also have the option to remove it in future if it is problematic.